enhancements/multi-arch/heterogeneous-architecture.md

```diff
@@ -0,0 +1,337 @@
+- Manifestlists support for the internal image-registry, Openshift builds (hence CI will not be supported for heterogeneous architecture payloads).
+- heterogeneous control plane.
+
+## Proposal
+
+The support for installation and upgrade of a heterogeneous architecture cluster and the associated components will be done in phases. The phases
+range from a bare bones no upgrade support installation to the full fledged installation and upgrade capability of a cluster on different topologies.
+The phases are listed below and details on some components of the phases are described below.
+
+Phase 0: Prototype
+- Aimed at delivering a prototype that users can experiment with.
+- Aimed at Azure/AWS with the primary supported architectures being x86 + ARM.
+- For supporting ARM nodes on Azure, [hyper-v support][BZ-1949614] is required which is landing in RHEL 8.6 and will be included in RHCOS for 4.11.
+- oc adm command changes to support manifestlist payloads.
+- Ability to assemble and publish manifest listed release payloads in a reproducible way.
+- Does not include upgrade support.
+- Is not intended for production.
+
+Phase 1: GA of basic support
+- Release pipeline needs to be setup to construct manifestlist based release payloads.
+- Upgrade support - CVO/Cincinnati support (heterogeneous -> heterogeneous payload upgrades only).
+- Does not include support for:
+  - image registry - internal image registry will not support manifest listed images.
+  - imagestreams importing full manifestlist content (will continue to import a single image when referencing manifestlists).
+
+Phase 2: Full basic functionality
+- Imagestream manifestlist support.
+- OLM/Operator ecosystem improvements.
+- oc debug (and other commands) support for compute architecture.
+- support toolbox.
+- support for Power and Z architectures.
+- Hypershift awareness.
+
+Phase 3: Quality of life improvements
+- First class install support - ability to create non-control-plane-architecture workers at install time and generate machineconfig
+and machineset manifests for the same.
+- Improve console's content filtering by making it architecture aware.
+- Possibly reduce mirroring costs by looking into things like sparse manifestlist.
+- Upgrade support for homogeneous payload -> heterogeneous payload and vice versa.
+
+### User Stories
+
+- As an enterprise customer looking to save compute resource cost, I want to be able to run part of my application workload on ARM instead of x86.
+- As the owner of an application that runs partially on x86 and partially on ARM, I want to be able to run the entire application within a
+single cluster so I do not need to pay the cost of higher latency between clusters, or worry about the HA-resilience of cross-cluster bridges.
+
+### API Extensions
+
+Upgrades to a heterogeneous payload will require an additional field either in the Infrastructure or the ClusterVersion resource to indicate that
+the user wants to upgrade to a heterogeneous payload.
```

Comment:
Will this get automatically set at install time when a user first installs a heterogeneous cluster?

what happens if it's not set on a heterogeneous cluster and an upgrade is attempted?

Reply:
i would expect CVO to automatically set it on first install based on whether its image is a manifestlist or not. Also on upgrades should CVO  also check the node architectures and automatically switch to heterogeneous payload if multiple arches are detected? i'll add that as an open question.

Reply:
This API extension is only required once we get to the phase 3 support for conveniently swapping between homogeneous and heterogeneous release images.

Reply:
Why can't we always use heterogeneous payload?

Reply:
for disconnected install cases we don't want the user to incur the additional cost of mirroring the extra arches if they only deal with homogeneous clusters.

Reply:
for disconnected installs, it's a higher mirroring cost to mirror a bunch of images for architectures you will never use. 

(our mirroring tools will likely grow the ability to filter out images you don't want, in conjunction with registries allowing sparse manifestlists, someday, but that's not the world we are building this for today)

for connected clusters, i'm not sure there's a reason a user would care other than, perhaps, easier management of CVE exposures in cases of a CVE that impacts an architecture they aren't actually using.

Reply:
updated proposal to reflect the migration and upgrade aspects.